spec/consensus/gossip/implementation.md

```diff
@@ -0,0 +1,212 @@
+If a message must be sent, it is enqueued for transmission using TCP and will either be delivered to the destination or the connection will be dropped.
+New connections reset the state of Communication layer wrt the newly connected node (in case it is a reconnection) and any messages previously sent (but possibly not delivered) will be resent if the conditions needed apply. If the conditions no longer apply, it means that the message has been superseded and need no be retransmitted.
+
+1. Every message added to BCast is enqueued or will be enqueued later.
+2. Every message enqueued will be delivered using TCP or the connection will break.
+
+## [GLOBAL-CAST.1]
+Messages delivered either cause the State to be advanced, causing the message to be superseded, or are added to the Communication layer internal state to be checked for matching conditions in the future.
+From the internal state it will affect the generation of new messages, which may have exactly the same contents of the original one or not, either way superseding the original.
+
+1. A received message is processed by the node, validator or not.
+2. The updated state causes the "same" message to be sent or a superseding one.
+
+
+
+
+
+
+
+
+
+
+
+### The inner workings
+
+CONS uses `broadcast` to send messages to all processes, but since the network may not be fully connected, communication happens at a **local** level, in which information is delivered to the neighbors of a node, and a **global level**, in which information is forwarded on to neighbors of neighbors and so on.
+While CONS is not aware of this distinction, GOSSIP is, since it is where the global level is implemented.
+
+* Messages
+    * proposal
+    * prevote
+    * precommit
+    * has-block-part
+    * has-vote
+* CONS state
+    * height[p]: Nat
+    * round[p]: Nat
+    * step[p]: {propose, prevote, precommit}
+    * decision[p]: List
+* GOSSIP state
+    * DMsgs[p]: set of messages delivered
+
+* Predicates
+
+> **TODO**
+> * Ne[p]: Neighbor set
+> * 
+
+
+
+**[PROV-GOSSIP-LOCAL_CAST.1]**   
+For any message $m1$ added to $\text{BMsgs}[p]$ at instant $t1$, let $\text{NePt1}$ be the value of $\text{Ne}[p]$ at time $t1$.
+For each process $q \in NePt1$, $m1$ will be delivered to $q$ at some point in time $t2 > t1$, or there exists a point in time $t3 > 1$ at which $q$ disconnects from $p$, or a message $m3$ is added to $\text{BMsgs}[p]$ at some instant $t3 > t1$ and $m3.\text{SSS}(m1)$.
+
+> **TODO**: Add permalink
+
+**[PROV-GOSSIP-GLOBAL_CAST.1]**   
+For every message received, either the message itself is forwarded or a superseding message is broadcast.
+
+> **TODO**: Add permalink
+
+
+#### Current implementations
+
+[REQ-GOSSIP-P2P-UNICAST]   
+- `Send(Envelope)`/`TrySend(Envelope)`
+    - Enqueue and forget. 
+    - Disconnection and reconnection causes drop from queues.
+    - Enqueuing may block for a while in `Send`, but not on `TrySend`
```

Comment:
Moreover, if enqueuing is not possible (because the remote peer is slow or crashed), both methods return `false`: `TrySend` immediately, `Send` after 10s.